x-pack/platform/plugins/shared/apm_sources_access/server/plugin.ts

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  type PluginInitializerContext,
+  type CoreSetup,
+  type CoreStart,
+  type Plugin,
+  type SavedObjectsClientContract,
+  type Logger,
+  SavedObjectsClient,
+} from '@kbn/core/server';
+import { OBSERVABILITY_APM_SOURCES_ACCESS_APM_SOURCES_ID } from '@kbn/management-settings-ids';
+import type { APMSourcesAccessConfig } from '../common/config_schema';
+
+import {
+  apmIndicesSavedObjectDefinition,
+  getApmIndicesSavedObject,
+} from './saved_objects/apm_indices';
+import { uiSettings } from '../common/ui_settings';
+
+/**
+ * APM Source setup services
+ */
+export type ApmSourcesAccessPluginSetup = ReturnType<ApmSourcesAccessPlugin['setup']>;
+/**
+ * APM Source start services
+ */
+export type ApmSourcesAccessPluginStart = ReturnType<ApmSourcesAccessPlugin['start']>;
+
+export class ApmSourcesAccessPlugin
+  implements Plugin<ApmSourcesAccessPluginSetup, ApmSourcesAccessPluginStart>
+{
+  public config: APMSourcesAccessConfig;
+  public logger: Logger;
+
+  constructor(initContext: PluginInitializerContext) {
+    this.config = initContext.config.get<APMSourcesAccessConfig>();
+    this.logger = initContext.logger.get();
+  }
+
+  getApmIndices = async (savedObjectsClient: SavedObjectsClientContract) => {
+    const apmIndicesFromSavedObject = await getApmIndicesSavedObject(savedObjectsClient);
+    return { ...this.config.indices, ...apmIndicesFromSavedObject };
+  };
+
+  /**
+   * Registers the saved object definition and ui settings
+   * for APM Sources.
+   */
+  public setup(core: CoreSetup) {
+    // register saved object
+    core.savedObjects.registerType(apmIndicesSavedObjectDefinition);
+    core.uiSettings.register({
+      [OBSERVABILITY_APM_SOURCES_ACCESS_APM_SOURCES_ID]: {
+        ...uiSettings[OBSERVABILITY_APM_SOURCES_ACCESS_APM_SOURCES_ID],
+        value: JSON.stringify(this.config.indices, null, 2),
+      },
+    });
+
+    // expose
+    return {
+      apmIndicesFromConfigFile: this.config.indices,
+      getApmIndices: this.getApmIndices,
+    };
+  }
+
+  /**
+   * Initialises the user value for APM Sources UI settings.
+   */
+  public start(core: CoreStart) {
+    const initSettings = async () => {
+      const soClient = core.savedObjects.createInternalRepository();
+      const indices = await this.getApmIndices(soClient);
+      const uiSettingsClient = core.uiSettings.asScopedToClient(new SavedObjectsClient(soClient));
+      await uiSettingsClient.set(
+        OBSERVABILITY_APM_SOURCES_ACCESS_APM_SOURCES_ID,
+        JSON.stringify(indices, null, 2)
+      );
```

Comment:
I'm confused here too. We're getting apm indices from a config to assign it to another setting?

Reply:
This `uiSetting` has both what is specified by the config file, and then what is specified by the user in the APM settings, which *override* the config file settings. So the resulting ui setting has the default `value` and the override `userValue` being set once all this has executed. With the discover page, I need to know the *end result* of the config, not just what is set on the `kibana.yml`.

Reply:
Yes, but doesn't `getApmIndices` do exactly that? What I don't understand is why we need to create this new `uiSetting` and not just call `getApmIndices`

Reply:
This is *on the server*, but the indices need to be exposed to the browser/clientside. The `savedObjects` client on the browser side is deprecated, so it was either this, or doing a whole HTTP API for it, if going by https://github.com/elastic/kibana/issues/149098.

Reply:
Have you considered, instead of creating this new setting, moving the `GET /internal/apm/data_view/index_pattern` into this plugin and calling this endpoint in `traces_context_services`?